### DISCUSSION TITLE
Is `FieldOffset` aliasing safe for reference types?

### DISCUSSION THREAD

**Discussion by colejohnson66** [dotnet/runtime#96321] [Q&A] [answered]
Suppose I am making a runtime for a language, and I want to use a discriminated union for storage of value types:

```cs
public readonly struct Value : IObject
{
    ulong _meta;     // discriminant and more
    IObject? _obj;   // null if a value type
    string? _string; // null if not a string
    bool _boolean;   // false if not a boolean
    long _integer;   // zero if not an integer
    double _float;   // zero if not a float

    // ...
}
```

The problem is: that's a lot of wasted space for unused fields. I figure I could use `StructLayoutAttribute` to lay them on top of each other and just be careful with access:

```cs
[StructLayout(LayoutKind.Explicit)]
public readonly struct Value : IObject
{
    [FieldOffset(0)]
    ulong _meta;

    [FieldOffset(8)]
    IObject? _obj;
    [FieldOffset(8)]
    string? _string;
    [FieldOffset(8)]
    bool _boolean;
    [FieldOffset(8)]
    long _integer;
    [FieldOffset(8)]
    double _float;

    // ...
}
```

However, now I've aliased value types on top of reference types. Even though I'd be able to tell when the bytes at offset eight are and aren't pointers (based on the metadata), the runtime/GC wouldn't. If `_integer` just so happens to be a "bad" enough value that it "looks like" a pointer, could the runtime get confused?

Is this safe in C#?

**Comment by alexrp** [comment, marked as answer]
It is not safe, and your example will result in a `TypeLoadException` when you try to use the `Value` type.
